scipy/stats/tests/test_stats.py

```diff
@@ -4480,6 +4494,67 @@ def test_constant_input(self, a, b, expected):
             f, p = stats.f_oneway(a, b)
             assert f, p == expected
 
+    def test_2d_inputs(self):
+        a = np.array([[1, 4, 3, 3],
+                      [2, 5, 3, 3],
+                      [3, 6, 3, 3],
+                      [2, 3, 3, 3],
+                      [1, 4, 3, 3]])
+        b = np.array([[3, 1, 5, 3],
+                      [4, 6, 5, 3],
+                      [4, 3, 5, 3],
+                      [1, 5, 5, 3],
+                      [5, 5, 5, 3],
+                      [2, 3, 5, 3],
+                      [8, 2, 5, 3],
+                      [2, 2, 5, 3]])
+        c = np.array([[4, 3, 4, 3],
+                      [4, 2, 4, 3],
+                      [5, 4, 4, 3],
+                      [5, 4, 4, 3]])
+
+        with assert_warns(stats.F_onewayConstantInputWarning):
+            f, p = stats.f_oneway(a, b, c)
+
+        # Verify the result computed with the 2d arrays matches the result
+        # from calling f_oneway individually on each slice.
+        for j in [0, 1]:
+            fj, pj = stats.f_oneway(a[:, j], b[:, j], c[:, j])
+            assert_allclose(f[j], fj, rtol=1e-14)
+            assert_allclose(p[j], pj, rtol=1e-14)
+        for j in [2, 3]:
+            with assert_warns(stats.F_onewayConstantInputWarning):
+                fj, pj = stats.f_oneway(a[:, j], b[:, j], c[:, j])
+                assert_equal(f[j], fj)
+                assert_equal(p[j], pj)
+
+    def test_length0_error(self):
+        # Require at least one value in each group.
+
+        with assert_warns(stats.F_onewayBadInputSizesWarning):
+            result = stats.f_oneway([1, 2, 3], [], [4, 5, 6, 7])
+            assert_equal(result, (np.nan, np.nan))
+
+        with assert_warns(stats.F_onewayBadInputSizesWarning):
+            ncols = 3
+            a = np.ones((4, ncols))
+            b = np.ones((0, ncols))
+            c = np.ones((5, ncols))
+            f, p = stats.f_oneway(a, b, c)
+            nans = np.full((ncols,), fill_value=np.nan)
+            assert_equal(f, nans)
+            assert_equal(p, nans)
+
+    def test_all_length_one(self):
```

Comment:
if you factor out into a parametrize annotation the 2 cases I mention above, this case could go in there as well and the test can be testing both cases, so there is a single test for the bad input warning.  That would make things a little more compact with no loss to readability-not to me anyway, let me know if you disagree.

Reply:
I decided to stick with separate tests.  Converting all three to a single parametrized test got a little ugly. In this case, I found the separate tests--with distinct test names--was clearer.

Reply:
yes, looking back at it now, I disagree with my prior comment, the names are very helpful here.